### PR TITLE
Add automatic redirect creation (lite)

### PR DESCRIPTION
**Pull request by ababic**
A lighter alternative to #7809 that avoids changes to `Page.get_url_parts()` for the sake of avoiding dummy requests (and the related refactoring that spurred)

Known limitation: Automatic redirect creation only works for the changed page and its descendants - it doesn't do anything directly with alternate site root paths / localised versions of pages

### CODE REVIEW

**Review by gasman — CHANGES_REQUESTED** [review]
Still getting my head around the redirect creation logic, but wanted to get this change request in for the time being.

File: wagtail/contrib/redirects/models.py [hunk 1]
```diff
@@ -71,16 +95,21 @@ def add_redirect(old_path, redirect_to=None, is_permanent=True):
...
         redirect.old_path = Redirect.normalise_path(old_path)
+        redirect.site = site
 
         # Check whether redirect to is string or Page
         if isinstance(redirect_to, Page):
             # Set redirect page
             redirect.redirect_page = redirect_to
+            # Set redirect page route
+            if isinstance(page_route_path, str):
+                redirect.redirect_page_route_path = Redirect.normalise_page_route(page_route_path)
```
[the remark this answers is not in the document]
Comment (ababic) [reply wagtail/contrib/redirects/models.py]: Thank you for sharing the tests for `RoutablePageMixin`. 100% agree that we should test this. I also noticed some other related things that needed fixing, so very happy!

File: wagtail/contrib/redirects/signal_handlers.py [hunk 2]
```diff
@@ -39,10 +39,9 @@ def autocreate_redirects_on_slug_change(
         return None
 
     # Determine sites to create redirects for
-    sites = Site.objects.exclude(root_page=instance).filter(id__in=[
-        option.site_id for option in instance._get_site_root_paths(request=instance)
-        if instance.url_path.startswith(option.root_path)
-    ])
+    sites = Site.objects.filter(id__in=[
+        option.site_id for option in instance._get_relevant_site_root_paths(request=instance)
```
Comment (zerolab) [inline wagtail/contrib/redirects/signal_handlers.py:43]: `TypeError: Page._get_relevant_site_root_paths() got an unexpected keyword argument 'request'`

Should be `cache_object=instance`, no?

Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:43]: Ah yes, of course. Late-night coding is rarely a good idea :) It's good that the CI failed at least... the test coverage is at least good :)

File: wagtail/core/utils.py [hunk 3]
```diff
@@ -356,3 +366,145 @@ def multigetattr(item, accessor):
...
+
+def get_dummy_request(path: str = "/", site: 'Site' = None) -> HttpRequest:
+    """
+    Return a simple ``HttpRequest`` instance that can be passed to
+    ``Page.get_url()`` and other methods to benefit from improved performance
+    when no real ``HttpRequest`` instance is available.
+
+    If ``site`` is provided, the ``HttpRequest`` is made to look like it came
+    from that Wagtail ``Site``.
+    """
```
Comment (gasman) [inline wagtail/core/utils.py:379]: Have been doing some code archaeology to figure out why we need to jump through these hoops, and whether we could massively simplify this by constructing a generic `HttpRequest` up-front in `create_redirects` and passing that as `cache_object`.

It comes down to a bad design decision in #4390 - the `request` argument to `get_url_parts` is being assigned a meaning that it was never meant to have [according to the API reference](https://docs.wagtail.io/en/stable/reference/pages/model_reference.html#wagtail.core.models.Page.get_url_parts) - namely, it's assuming that we're building a URL for the purpose of rendering it into a page template, and `request` is the request that initiated that template response.

The one and only time when it comes into play is when there are multiple sites pointing to the same root page, so rather than selecting the deepest-rooted one as usual, it will break the tie by preferring the one that matches `request`'s site.

(I'm discounting the possibility that a page model might override `get_url_parts` to do its own custom `request`-dependent logic, because that clearly goes against the instructions in the `get_url_parts` docs.)

So, unwinding a bit: the only time `_page_urls_for_sites` will ever return multiple results is if the `get_url_parts` result varies on request, and that only happens if this 'multiple sites pointing to the same root' situation is in effect. And in that case, the multiple results returned from `_page_urls_for_sites` will be something of the form

    {
        (first_aliased_site, '/foo', '/bar'),
        (second_aliased_site, '/foo', '/bar'),
    }

where first_aliased_site and second_aliased_site are the multiple sites pointing to the same root, and the other fields are identical.

In short: we only need these specially-crafted request objects to deal with one annoying edge case, and in that edge case, it should be fairly easy to derive the intended results from a single `get_url_parts` call with a generic request. However, by now I think the reasoning to get there is more complicated than the code we've got here, so let's stick with the current version for now - I will personally take on the technical debt and deal with that in a future PR :-)

**Review by gasman — CHANGES_REQUESTED** [review]
Two small points, and a lengthy ramble that can be disregarded for the purpose of getting this merged :-)

File: wagtail/contrib/redirects/signal_handlers.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+) -> None:
+    if not getattr(settings, "WAGTAILREDIRECTS_AUTO_CREATE", True):
+        return None
+
+    if url_path_after == url_path_before:
+        # Redirects are not needed for a page 'reorder'
+        return None
+
+    # NB: create_redirects expects specific instances
+    instance = instance.specific
```
[the remark this answers is not in the document]
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:63]: I will double check, but I believe the `page_slug_changed` signal provides specific pages already... I did things that way, because it would be very easy to do `old_instance.specific` in a handler and lose all of the original values.

File: wagtail/contrib/redirects/models.py [hunk 5]
```diff
@@ -121,9 +150,32 @@ def normalise_path(url):
...
+    @staticmethod
+    def normalise_page_route_path(url):
+        # Strip whitespace
+        url = url.strip()
+        if not url:
+            return ""
+
+        # Extract the path from the rest of the value
+        url_parsed = urlparse(url)
+        path = url_parsed[2]
```
[the remark this answers is not in the document]
Comment (ababic) [reply wagtail/contrib/redirects/models.py]: Thanks @gasman. Agree that is more readable. Updated in 6d57206

File: wagtail/contrib/redirects/signal_handlers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ababic) [reply wagtail/contrib/redirects/signal_handlers.py:66]: @gasman Have added a couple of comments in 1c47145 to help clarify this

**Review by gasman — APPROVED** [review]

**Comment by ababic** [comment]
@gasman Have just rebased for you :)
